### ISSUE TITLE
Pinebook Pro support

### ISSUE THREAD

**Issue by jcgruenhage** [void-linux/void-mklive#105]
This issue tracks progress towards a usable voidlinux image for the Pinebook Pro. Currently, we can already build a working image

 - [x] packaging
   - [x] pinebookpro-kernel (https://github.com/void-linux/void-packages/pull/19293)
   - [x] pinebookpro-uboot (https://github.com/void-linux/void-packages/pull/19196)
   - [x] pinebookpro-base (https://github.com/void-linux/void-packages/pull/19216)
   - [x] pinebookpro-firmware (https://github.com/void-linux/void-packages/pull/19197)
 - [x] add support to scripts in this repo (#109)
   - [x] set_target_arch_from_platform in lib.sh.in
   - [x] mkplatformfs
   - [x] mkimage
 - [x] fix broken things in the scripts here
   - [x] make sure /tmp is mounted correctly (#110)
   - [x] fix post-install scripts on cross builds (#111)
 - [ ] Images available on https://alpha.de.repo.voidlinux.org/live/

How to read the list: A checked box means that the work is done and merged, an unchecked box with a PR linked means that this is basically done, but not merged yet.

**Comment by Vaelatern** [comment]
Can we try to use the generic uboot and kernel for this? I believe upstream uboot can run this machine

**Comment by jcgruenhage** [comment]
If that's possible, absolutely. The kernel still needs a few patches, but is very close to mainline

**Comment by jcgruenhage** [comment]
If that's possible, absolutely. The kernel still needs a few patches, but is very close to mainline

**Comment by Duncaen** [comment]
upstream uboot supporting the device does not allow to build a "generic" uboot.

**Comment by xtraeme** [comment]
No hardware, but I can help with packaging stuff... just FYI :-)

**Comment by jcgruenhage** [comment]
@xtraeme I do have the hardware here and am willing to test stuff, but I have no experience packaging things like the linux kernel, so your help is appreciated!

A lot of the work done by the manjaro people can be used, like https://gitlab.manjaro.org/manjaro-arm/packages/core/linux-pinebookpro/tree/master for example, so it's mostly a job of "translating" stuff from their PKGBUILDs to our templates

**Comment by houki0** [comment]
I've already been playing around with a simple rootfs install by replacing the root partition of the default debian image, so it's running their kernel and uboot right now. It booted just fine out of the box, but I did have to copy the wifi firmware to get wifi to work.

Maybe we could make a pbp-firmware package like the one for the rpi3?

That said, I should be able to test most things, but I'm not great at packaging, and stuff like configuring kernels and uboot is probably beyond me.

**Comment by BenEvolent333** [comment]
I don't have any packaging knowledge, but I have a pbp and would be willing to offer my services to get void running on it.

**Comment by jcgruenhage** [comment]
I think we don't lack testers, we mostly lack someone with packaging experience and time who is interested in working on this. I have a bit of experience and am certainly interested, but my schedule is pretty packed right now..

**Comment by renatoaguiar** [comment]
I can do the packaging. I'm also pretty busy right now, but I hope to find some time to work on that over the weekend.

**Comment by sunnaryt** [comment]
would really like to see this happen, i only have packaging experience with arch though. i have a lot of free time coming up, like 2 months free time for reasons and PBP is my favorite new toy, i will learn and help maintain if someone points me in the right direction.

**Comment by renatoaguiar** [comment]
I just pushed WIP to 'pbp' branch on my forks:
  - https://github.com/renatoaguiar/void-packages/tree/pbp
    - pinebookpro-kernel (build works, boot not yet tested)
    - pinebookpro-base
  - https://github.com/renatoaguiar/void-mklive/tree/pbp
    - add arch detection for pinebookpro
    - add pinebookpro to mkplatformfs

I'm now working on packaging pinebookpro-uboot.

**Comment by q66** [comment]
you know, there is already https://github.com/void-linux/void-packages/pull/17672

**Comment by q66** [comment]
Also, we don't want a 4.4 kernel, it won't work with panfrost and the proprietary libmali which works with 4.4 doesn't work with modern userlands (and if you hack around it it still works poorly)

**Comment by renatoaguiar** [comment]
> Also, we don't want a 4.4 kernel, it won't work with panfrost and the proprietary libmali which works with 4.4 doesn't work with modern userlands (and if you hack around it it still works poorly)

From what I read so far, mainline kernel still doesn't support some features, e.g. suspend/resume. My plan is to initially use same version as official (factory) build and upgrade to mainline kernel as soon as it becomes feature complete.

**Comment by q66** [comment]
Suspend/resume may not work, but accelerated graphics not working is worse.

You can use s2idle for suspend for now, which works.

**Comment by houki0** [comment]
Assuming the kernel @renatoaguiar packaged is the same one I copied from the default debian (I can confirm they are both version 4.4.202):

I am currently unable to launch any DE or WM both for wayland and x11 (I tested XFCE, KDE on both and Sway). I believe it is an issue with dbus because it seems NetworkManager doesn't work either (I normally use connman so I hadn't noticed before), but runit reports dbus as up, and I can't find any errors with socklog.

I haven't tried a 5.4 or 5.5rc kernel from Manjaro yet because I haven't got those to boot yet, and I don't have a working UART cable to see what's wrong.

In any case, it might be worth looking into.

**Comment by q66** [comment]
Wayland won't work on 4.4, except weston fbdev. Sway needs drm, and for that you need libmali, and libmali does not expose enough of libgbm to have weston-drm/sway working. Out of curiosity, I set up proprietary libmali and went to make a preload library to stub out the missing libgbm calls, and got far enough to launch weston and sway with drm, but EGL still does not work. In X11, you will not get acceleration either, since GLAMOR requires EGL, which does not work. So yeah, unaccelerated fbdev or at most drm with llvmpipe is best you will get with the 4.4 kernel. Use 5.5-rc7 or 5.4 from the manjaro tree and panfrost if you want accel of any kind.

**Comment by houki0** [comment]
Ah, that probably explains why KDE was complaining about $DISPLAY not being set. I didn't know about fbdev. I'm learning new things by the minute.

In that case, I'll try to get a more recent kernel to run after I fix my UART cable.

**Comment by jcgruenhage** [comment]
I updated the top post with the WIP stuff that happened so far. I think going with a 4.4 kernel doesn't make sense, since mainline is so close to being feature complete. I'm using the manjaro build at the moment and it works definitely good enough for me.

**Comment by CameronNemo** [comment]
Firmware for the mainline kernel is here: https://gitlab.manjaro.org/tsys/pinebook-firmware/ .

I pushed 5.5 to the PR linked in the top comment, but have yet to test it.

**Comment by q66** [comment]
I'm using 5.5-rc7-panfrost-fixes branch and it works ok

**Comment by renatoaguiar** [comment]
I was checking the latest news and It turns out @q66 was right from the beginning :)

Mainline kernel support is progressing faster than I initially thought, so I'll be focusing now on verifying the existing kernel package from @CameronNemo and packaging whatever is still missing.

**Comment by houki0** [comment]
I just finished compiling @CameronNemo 's kernel, and looking at the commit referenced in the template, it is the head of the v5.5 branch of manjaro's kernel.

That said, it boots just fine and sway seems to work, although scrolling in Wayland-native firefox is a bit laggy. Is there any way for Wayland to check if it's using software rendering? As far as I can tell glxinfo is only accurate for X11 and XWayland. XWayland firefox works just fine and mesa is installed.

I'll test some other DE's later this week.

**Comment by renatoaguiar** [comment]
I just added pinebookpro-uboot to https://github.com/renatoaguiar/void-packages/tree/pbp. It still needs some cleanup, but I was good enough to generate a bootable image from https://github.com/renatoaguiar/void-mklive/tree/pbp.

@CameronNemo could you rename your linux-pinebook-pro to pinebookpro-kernel, so it matches naming convention used on kernels for other platforms?

**Comment by renatoaguiar** [comment]
pinebookpro-uboot: https://github.com/void-linux/void-packages/pull/19196
pinebookpro-firmware: https://github.com/void-linux/void-packages/pull/19197

**Comment by anjandev** [comment]
I tried @renatoaguiar's build scripts. Just for anyone else that wants to try this:

clone https://github.com/renatoaguiar/void-packages and run  `git checkout pbp && ./xbps-src binary-bootstrap && ./xbps-src -a aarch64 pkg pinebookpro-base`.

Run  `./xbps-src -a aarch64 pkg wpa_supplicant` cause the default rootfs doesnt come with a wireless...

Then, clone https://github.com/renatoaguiar/void-mklive and run `git checkout pbp && make`.
The rootfs off void's website has an outdated xbps that won't let us run `mkplatform.sh`. It's ok because we can generate one with `./mkrootfs.sh -o aarch64latest aarch64`.

Next, we run `sudo ./mkplatformfs.sh -r ../void-packages/host
binpkgs/pbp/ -p 'pinebookpro-base wpa_supplicant' pinebookpro aarch64latest`.

Finally, to generate the img, I ran: "sudo ./mkimage.sh void-pinebookpro-PLATFORMFS-20200216.tar.xz"

I burned it to my eemc and it works! It's 4 am and I have been up since 7am so I will do further testing in the morning. Thanks for your work. =)

**Comment by renatoaguiar** [comment]
> I tried @renatoaguiar's build scripts. Just for anyone else that wants to try this:

Thanks for trying that out and posting instructions :)

> Run `./xbps-src -a aarch64 pkg wpa_supplicant` cause the default rootfs doesnt come with a wireless...

wpa_supplicant for aarch64 is available in official repos. You don't need to build it locally.

> Next, we run `sudo ./mkplatformfs.sh -r ../void-packages/host binpkgs/pbp/ -p 'pinebookpro-base wpa_supplicant' pinebookpro aarch64latest`.

I didn't have to specify any extra packages in mkplatformfs. At least pinebookpro-base and wpa_supplicant are being installed automatically as part of pinebookpro platform, so you shouldn't need to list them in '-p'.
 
> I burned it to my eemc and it works! It's 4 am and I have been up since 7am so I will do further testing in the morning. Thanks for your work. =)

I haven't tried booting from emmc yet. Good to know that it works too :)

**Comment by jcgruenhage** [comment]
So, to summarize: We have WIP stuff for everything now, we can use that to actually build images that work too, we just need to get the stuff merged now. Some thing have already gotten PRs, for the rest that still needs to be done. I updated the issue description, have I missed something there? Do we have something that's merged already?

**Comment by renatoaguiar** [comment]
void-mklive: https://github.com/void-linux/void-mklive/pull/109

**Comment by renatoaguiar** [comment]
pinebookpro-base: https://github.com/void-linux/void-packages/pull/19216

**Comment by anjandev** [comment]
@renatoaguiar So I installed void on the emmc but I changed my fstab in a way that it void failed to boot. Unfortunately, this failure to boot is not visible to the user. I looked to flash manjaro to an sd card and boot from it to fix the issue. However, the uboot on void doesnt prioritize booting from the sdcard before emmc like on most distros for pinebook pro.

This is an important bug that needs to be fixed before these pull requests can be merged. 

I fixed my issue by following the advice here: https://forum.pine64.org/showthread.php?tid=8790

**Comment by jcgruenhage** [comment]
I've built an image yesterday and I can't get it to boot from eMMC or from the sdcard..

**Comment by KeepBotting** [comment]
I have attempted to create some bootable media for Void on the Pinebook Pro using the packages and instructions present in this issue.
Near as I can tell, the image was created properly, but I have been unable to boot from SD card.

Here are my steps. I ran these commands on a 64-bit x86 server running Void glibc.

```
git clone https://github.com/renatoaguiar/void-packages void-packages-pbp && cd void-packages-pbp
git checkout pbp && ./xbps-src binary-bootstrap
./xbps-src -a aarch64 -j8 pkg pinebookpro-base
```

The build for `pinebookpro-base` (which includes the kernel) is using eight makejobs (`-j8`), so scale this according to your hardware.

```
cd ../
git clone https://github.com/renatoaguiar/void-mklive mklive-pbp && cd mklive-pbp
git checkout pbp && make
sudo ./mkrootfs.sh -o aarch64latest aarch64
```

However, the previous command errored with `FATAL: qemu-aarch64-static binary is missing in your system, exiting.`

So I installed `qemu-user-static` and then attempted rebuilding the rootfs again. This time it was successful.

Continuing, the only extra package I can see myself wanting straight away is `wifish` as this is my preferred way to manage wireless network associations from the command-line.

```
sudo ./mkplatformfs.sh -r ../void-packages-pbp/hostdir/binpkgs/pbp/ -p 'wifish' pinebookpro aarch64latest
sudo ./mkimage.sh void-pinebookpro-PLATFORMFS-*.tar.xz
```

Finally, I decompressed the image and flashed it to a SD card.

```
unxz void-pinebookpro-*.img.xz
# be careful with dd, as always
sudo dd status=progress if=void-pinebookpro-*.img of=/dev/sdX
```

After flashing, I inserted the SD card into my Pinebook Pro, but it failed to boot. Instead, the default Debian (on eMMC) booted.
I added the `boot` flag to the Void partition on the SD card and tried again, but this did not have any effect.

My Pinebook Pro still has the default Debian image present on eMMC, so the bootloader is (presumably?) untouched. 
I have not tried writing the Void image to eMMC just yet. I am not quite comfortable with u-boot (I have only used GRUB and rEFInd), and am hesitant to wipe out the eMMC, since that is where u-boot lives. I would rather not have to open the laptop to recover.

What should I do differently?

Here is the partition layout for my Manjaro ARM SD card (which boots properly): https://branon.me/files/NjVkM2E.png
And here is Void's SD card: https://branon.me/files/NGJkNWR.png

**Comment by jcgruenhage** [comment]
I just rebuilt it and it still doesn't work. OOI, can we gather info on who here 1) had success and 2) what revision they have? Maybe all the people who had success have devices from the same batch and there's some weirdness going on with newer/older revisions. I had no success, from both eMMC and sdcard, on a device from the first ANSI batch

**Comment by renatoaguiar** [comment]
@jcgruenhage @KeepBotting could you try again using https://github.com/renatoaguiar/void-mklive/tree/pbp-local?

It reintroduces some local changes I had in pbp branch before opening the PR.

**Comment by KeepBotting** [comment]
I am using an ANSI model. I am unsure how to tell which batch it is from. It was ordered on 27th December and arrived on 20th January 2020.

**Comment by KeepBotting** [comment]
Odd, using `pbp-local` branch fails with an error on this command: `sudo ./mkrootfs.sh -o aarch64latest aarch64`

```
Reconfiguring packages for aarch64 ...
Running xbps-reconfigure --rootdir /tmp/tmp.J90tU0vw5u base-files
base-files: configuring ...
ERROR: base-files: [configure] INSTALL script failed to execute the post ACTION: No such file or directory
Failed to reconfigure `base-files': No such file or directory
chroot: failed to run command ‘sh’: No such file or directory
chroot: failed to run command ‘sh’: No such file or directory
Setting the default root password ('voidlinux')
chroot: failed to run command ‘sh’: No such file or directory
chroot: failed to run command ‘sh’: No such file or directory
FATAL: Could not set default credentials
```

**Comment by renatoaguiar** [comment]
@KeepBotting you probably have stale entries on your binfmt.

Try running `$ echo -1 | sudo tee /proc/sys/fs/binfmt_misc/status` before the build to remove all entries.

**Comment by Duncaen** [comment]
Try with this patch https://patch-diff.githubusercontent.com/raw/void-linux/void-mklive/pull/110.patch

**Comment by KeepBotting** [comment]
@renatoaguiar Interesting, that did indeed solve the error. I have never interacted with binfmt_misc before, but I gather it's used because I am cross-compiling for aarch64 on an x86 host. The patch provided by @Duncaen should fix the configure scripts so manual intervention is not required.

I'm happy to say I was able to successfully boot from the Void SD card after flashing this new image. The process used was identical to my original comment above, with the exception of using `pbp-local` branch of void-mklive.

Wi-Fi works properly of course, and I was able to install X11 and Mesa to achieve hardware acceleration with the Panfrost driver.

Opening Firefox's filechooser dialog causes X to segfault and crash, however. Not sure what that's about.

**Comment by q66** [comment]
panfrost is rather unstable still and has issues with GLAMOR, you should run mesa from git for best experience

**Comment by renatoaguiar** [comment]
I pushed another PR https://github.com/void-linux/void-mklive/pull/111 to void-mklive fixing the execution of post install scripts on cross builds. That is also required to successfully create an image from https://github.com/void-linux/void-mklive/pull/109.

I cleaned https://github.com/renatoaguiar/void-mklive/tree/pbp-local up by keeping only changes from opened PRs, so people should still be able to build images from it.

**Comment by anjandev** [comment]
It may be worthwhile to start documenting the pinebook pro. I ran into keyboard mapping on my iso keyboard and documented it here: https://wiki.voidlinux.org/Pinebook_Pro. Please let me know if you would like me to add anything else to the wiki.

I am running the original img without issues for the last 24 hours. It's been great!

**Comment by jcgruenhage** [comment]
I've built an image using the `pbp-local` branch rebased on upstream master, this image boots (from eMMC), but once I install a bit of stuff and resize the partition, it doesn't boot anymore. I'll investigate a bit more *what exactly* causes it to not boot anymore, but without an external eMMC flasher and a uart adapter, this is a bit hard.

**Comment by BenEvolent333** [comment]
I have booted it once since I installed it, so far everything appears to be working perfectly on my pbp. Fwiw, I got the batch that shipped January 2020, which I believe is the second batch, and I flashed to SD, not to emmc. I installed Xorg and the MATE desktop. First thing I notice is that every time I start X, the mouse is jittery for a few seconds, then moves normally. Also, what method would we use to upgrade the pbp base image (I.e. kernel and uboot) from the installed computer? Do we clone `https://github.com/renatoaguiar/void-packages` and use that particular xbps source?

**Comment by renatoaguiar** [comment]
@BenEvolent333 since you already have it working, I'd just wait until packages are available in the official mirrors. pinebookpro-uboot and pinebookpro-firmware are already merged, others are coming soon.

**Comment by renatoaguiar** [comment]
pinebookpro-kernel: https://github.com/void-linux/void-packages/pull/19293

I also fixed musl builds, so now it is also possible to make images for pinebookpro-musl.

**Comment by BenEvolent333** [comment]
I ran into a problem with Xorg, after I installed panfrost and then logged out of mate (using lightdm), where xorg no longer loads properly. I uninstalled panfrost, but the issue persisted. I attempted to load mate, lightdm, KDE plasma, abiword, and MATE terminal using xinit, of which only mate-terminal loads successfully. Somehow I also deleted the services directory, so I have to re-flash the image anyways, but I will update if I can pinpoint a cause.

**Comment by renatoaguiar** [comment]
Now https://github.com/void-linux/void-mklive/pull/112 is also needed because older versions of some packages are already available in official repos. I updated the pbp-local branch with that PR.

**Comment by xtraeme** [comment]
yes, it was my mistake for not thinking about 0.0.<date>... this decision was made long ago but I didn't remember... anyway it's almost complete. I merged all relevant bits, except the ones for qemu. Still unsure what decision to take :-)

**Comment by xtraeme** [comment]
I've just built a pinebookpro-musl image ready for testing. void-mklive contains all changes now.

**Comment by xtraeme** [comment]
@renatoaguiar can you confirm a void image for pinebookpro works now?

**Comment by xtraeme** [comment]
With all up-to-date components and repos.

**Comment by xtraeme** [comment]
would be nice to create a void news post. This device is so useful that we should mention about its support.... imho

**Comment by xtraeme** [comment]
also document mesa requirements for OpenGL or OpenGL/ES, whatever...

**Comment by jcgruenhage** [comment]
> can you confirm a void image for pinebookpro works now?
@xtraeme you mean when building from the master branch of void-mklive, without building any packages from source, right?

**Comment by xtraeme** [comment]
@jcgruenhage if all pkgs you target are there you don't need to build anything, this is the power or mkrootfs/mkplatform.

This is has been designed this way to make it build once via xbps-src and the use mkrootfs/mkplatforms to generate your image.

**Comment by xtraeme** [comment]
if we would use a generic linux ARM kernel with all dts files (like NetBSD) it would work in lots of devices... but I can't test. Open for hw testers ;.)

**Comment by KeepBotting** [comment]
Nice to hear about musl images! Going to try that out ASAP.

News post sounds like a great idea. Once support is finalized we will need to add an entry to the Pinebook Pro wiki page here: https://wiki.pine64.org/index.php/Pinebook_Pro_Software_Release

**Comment by xtraeme** [comment]
I'd like to know about cubieboard, rpi[1-4] users, and other ARM devices. I think 5.5 is already so powerful to make your device useful. If enough testers are there for these platforms I'll make something for you test. A generic linux kernel.

**Comment by xtraeme** [comment]
Just let me know, we can do better than NetBSD, note they are using older .dts files, so why not?

**Comment by xtraeme** [comment]
The linux kernel is so powerful, we just need to be smart enough to update our confs.

**Comment by xtraeme** [comment]
I'm pretty sure a generic ARM linux kernel would boot in 70% of devices targetted... this is it. There's device-tree for a reason, we just need to use it... like NetBSD.

**Comment by jcgruenhage** [comment]
`sudo ./mkplatformfs.sh pinebookpro <aarch64 rootfs>` should work, right? This currently gives me `FATAL: mkplatformfs.sh: invalid platform!`

**Comment by jcgruenhage** [comment]
`sudo ./mkrootfs.sh -o aarch64latest aarch64` is what I used to generate the rootfs

**Comment by xtraeme** [comment]
if you've got a rare ARM device (ARMv5+) contact me and I'll try to make a generic linux kernel to make it work in all known supported devices.

**Comment by xtraeme** [comment]
@jcgruenhage git pull, I merged the pinebookpro stuff a while ago

**Comment by jcgruenhage** [comment]
I did pull right before building

**Comment by xtraeme** [comment]
mkrootfs expects arch or arch-musl... i,e `./mkrootfs ... aarch64-musl`

**Comment by renatoaguiar** [comment]
> @renatoaguiar can you confirm a void image for pinebookpro works now?

It worked fine, but I had to build pinebookpro-base from void-packages (master) because it is not yet available in mirrors. I'm trying musl build now.

**Comment by xtraeme** [comment]
@renatoaguiar that's because they are busy building other stuff in queue

**Comment by xtraeme** [comment]
if you build this locally and use -r it works as expected, I just tested (glad it works because it's my code)

**Comment by jcgruenhage** [comment]
Don't mind my previous report, I pulled master from @renatoaguiar's fork... Trying again now with the upstream master branch.

**Comment by renatoaguiar** [comment]
I tried glibc and musl images from sdcard and they both worked :)

Can someone try booting from emmc?

**Comment by jcgruenhage** [comment]
building the image right now, will try eMMC boot after that

**Comment by jcgruenhage** [comment]
Random thought: We should make the uboot prefer the sdcard over the eMMC before publishing this as done to the community, as booting from the sdcard after flashing this image currently requires toggling the eMMC off using a hardware switch inside the chassis of the PBP.

**Comment by xtraeme** [comment]
I agree with @jcgruenhage. The image should prefer sdcard and then eMMC.

**Comment by jcgruenhage** [comment]
The image I built does not boot from eMMC it seems.

**Comment by xtraeme** [comment]
Please verify, we'd like to have this behaviour by default.

**Comment by jcgruenhage** [comment]
I mean it doesn't boot at all, not that it prefers the sdcard

**Comment by renatoaguiar** [comment]
I've never tried, but @anjandev tried before and it worked at some point.

**Comment by jcgruenhage** [comment]
I also had eMMC boot work at some point..
https://cloudflare-ipfs.com/ipfs/QmRCcbSic46xgtYLZEj2SFe97HMYF88CKWKj3MNDpCN721 is the image I built, can someone test that on an sdcard to see whether it's generally broken?

**Comment by jcgruenhage** [comment]
@renatoaguiar could you upload your known-to-work images so that I can try those on my PBP?

**Comment by xtraeme** [comment]
We need to test the generated ones.

**Comment by xtraeme** [comment]
I've got glibc and musl generated ones, I can upload them to everywhere...

**Comment by jcgruenhage** [comment]
If you send me download links, I will try them from eMMC

**Comment by renatoaguiar** [comment]
The ones I tested earlier with sdcard are available at https://volans.renatoaguiar.net/pub/

**Comment by jcgruenhage** [comment]
seems these neither boot from sd nor emmc for me..

**Comment by jcgruenhage** [comment]
My guess is that they do *some* booting, but the display just doesn't work for some reason. Debugging that isn't really possible without their 4 pin aux to usb uart thingy.

**Comment by renatoaguiar** [comment]
It looks like my emmc reader and console cable are coming tomorrow, so I can try and debug booting from emmc.

**Comment by CameronNemo** [comment]
@renatoaguiar I have updated my pinebookpro-kernel PR, and I tested the firmware package.

**Comment by CameronNemo** [comment]
I opened a clean, fresh PR. Want to keep the old branch handy in case anyone wants the old packages. https://github.com/void-linux/void-packages/pull/19336

**Comment by sunnaryt** [comment]
that image works for me, until i resize it. i believe it might be an issue with the way uboot on pbp detects a bootable device, but i'm just speculating. i don't have a console cable either

**Comment by renatoaguiar** [comment]
@sunnaryt boot partition is separate, so I guess it is booting fine. Maybe the root= argument passed to the kernel is wrong after resizing. How are you resizing the partition? Any chances the PARTUUID is changing? Did you change anything in /etc/fstab?

**Comment by BenEvolent333** [comment]
Mine continues to wok after resizing the partition, hmm

**Comment by sunnaryt** [comment]
fresh copy:
-boots first time
-boots 2nd time (to make sure something isn't changing just by booting it)
-did nothing else
-record uuid
-resize via gnome-disks from other linux install (boot untouched)
-uuids are the same
-boot fails
-uuids are still the same

i did nothing besides resize....hmmmm

maybe a different resize method? what do you use @BenEvolent333

**Comment by Skirmisher** [comment]
Manjaro includes a [pinebookpro-post-install](https://gitlab.manjaro.org/manjaro-arm/packages/community/pinebookpro-post-install) package that adds a couple supplementary files that could be included in `pinebookpro-base`:
- `/var/lib/alsa/asound.state` – ALSA mixer configuration
- `/etc/udev/hwdb.d/10-usb-kbd.hwdb` – mappings for sleep and brightness keys
They also edit the logind config to force s2idle, but that can also be forced with a `mem_sleep_default=s2idle` on the kernel command line.  As an aside, the Manjaro kernel supports suspend-to-RAM correctly, the issue lies with the ATF (ARM trust firmware) in mainline U-boot (it works with Rockchip's fork). (Also, the kernel is not truly 5.5.0, but rebased on master shortly after 5.5.0's release for some reason...)

Additionally, the RockPro64 is essentially a subset of the PBP hardware—same SoC and a number of similar peripherals. I've booted an unmodified PBP kernel on the RP64 before, so I could see a `rockpro64-kernel` vpkg of some sort existing, or maybe the main package being named `rk3399-kernel`, specifying PBP and RP64 in the description. Will need testing, of course, and something to determine which device tree to point the bootloader at.

Speaking of kernel packaging, from past experience I'm a little annoyed with how the device-specific kernel packages function. There's no old kernel retention/`vkpurge` support even though it seems quite possible, and it's ugly when packages (e.g. dkms modules) pull in the mainline kernel headers when the system doesn't need them (and then dkms wastes time trying to build for the mainline kver). `ignorepkg` lets you work around the latter, but it isn't a real solution. I would like to try to improve this and other issues (e.g. naming consistency) in the future to streamline the experience of using Void on ARM devices.

**Comment by renatoaguiar** [comment]
> Manjaro includes a [pinebookpro-post-install](https://gitlab.manjaro.org/manjaro-arm/packages/community/pinebookpro-post-install) package that adds a couple supplementary files that could be included in `pinebookpro-base`

@Skirmisher thanks for pointing that out. I'll take a look and update pinebookpro-base as needed.

**Comment by BenEvolent333** [comment]
@sunnaryt I used cfdisk, and then resize2fs, from my main laptop before booting from the SD. I also used resize2fs from the pbp while it was running before I had to reinstall because of xorg, and it rebooted just fine.

**Comment by BenEvolent333** [comment]
I just updated my void install, and installed plasma, and it would seem that no audio devices or Bluetooth adapters were found.

**Comment by BenEvolent333** [comment]
@Skirmisher  are you saying the issue with suspend-to-RAM could be fixed with using rockchip's fork of uboot?

**Comment by houki0** [comment]
I don't think mkimage.sh includes a uboot in the image.

cfdisk doesn't report ~64mb of empty space preceding any partitions, the same as in @KeepBotting 's post:
> Here is the partition layout for my Manjaro ARM SD card (which boots properly): https://branon.me/files/NjVkM2E.png
> And here is Void's SD card: https://branon.me/files/NGJkNWR.png

**I think that's why it doesn't boot from eMMC yet.**

Here's the output my UART gives me if I flash the image to eMMC:

```U-Boot TPL 2020.01-2-g365495a329 (Feb 20 2020 - 04:17:35)
Channel 0: LPDDR4, 50MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS1 Row=15 CS=2 Die BW=16 Size=2048MB
Channel 1: LPDDR4, 50MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS1 Row=15 CS=2 Die BW=16 Size=2048MB
256B stride
256B stride
lpddr4_set_rate: change freq to 400000000 mhz 0, 1
lpddr4_set_rate: change freq to 800000000 mhz 1, 0
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.01-2-g365495a329 (Feb 20 2020 - 04:17:35 +0000)
Trying to boot from MMC2
mmc_load_image_raw_sector: mmc block read error
Trying to boot from MMC2
mmc_load_image_raw_sector: mmc block read error
Trying to boot from MMC1
mmc_load_image_raw_sector: mmc block read error
SPL: failed to boot from all boot devices
### ERROR ### Please RESET the board ###```

**Comment by renatoaguiar** [comment]
@houki0 that's a very good catch. uboot is actually being installed, but the '/boot' partition is overlapping with it. This should cause it to be randomly overwritten at runtime while operating on '/boot' filesystem.

I pushed a fix for that to pbp-local branch of my void-mklive fork (https://github.com/renatoaguiar/void-mklive/tree/pbp-local). @jcgruenhage @sunnaryt, could you guys try it out and check if that fixes your boot/resize problems?

I also pushed new images with that fix at https://volans.renatoaguiar.net/pub/

**Comment by Skirmisher** [comment]
> @Skirmisher are you saying the issue with suspend-to-RAM could be fixed with using rockchip's fork of uboot?

That's correct; the Manjaro kernel developer was using that when patching suspend into the kernel, not realizing that it didn't work on upstream. (Manjaro was previously using Rockchip's fork of 2017.09, from [this tree](https://github.com/mrfixit2001/rockchip-u-boot). The PKGBUILD for that can be found [here](https://gitlab.manjaro.org/manjaro-arm/packages/core/uboot-pinebookpro/blob/930614c2003a8e05bb1abb48b31e9ff2140eb560/PKGBUILD).)

I don't know how long it'll take for the ATF to get fixed, but so far Manjaro hasn't reverted to the Rockchip version. But they require users to install it manually, so many people are still using the original one anyway.

**Comment by houki0** [comment]
@renatoaguiar 
Your new image boots perfectly and lightning fast!
There's a couple weird/useful things with it though:

- I can actually see part of the kernel log on the screen.
- When it boots, it first displays a red power light, which then turns green after a few seconds.

Here's the boot log from the UART:

```U-Boot TPL 2020.01-2-g365495a329 (Feb 20 2020 - 04:17:35)
Channel 0: LPDDR4, 50MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS1 Row=15 CS=2 Die BW=16 Size=2048MB
Channel 1: LPDDR4, 50MHz
BW=32 Col=10 Bk=8 CS0 Row=15 CS1 Row=15 CS=2 Die BW=16 Size=2048MB
256B stride
256B stride
lpddr4_set_rate: change freq to 400000000 mhz 0, 1
lpddr4_set_rate: change freq to 800000000 mhz 1, 0
Trying to boot from BOOTROM
Returning to boot ROM...

U-Boot SPL 2020.01-2-g365495a329 (Feb 20 2020 - 04:17:35 +0000)
Trying to boot from MMC2


U-Boot 2020.01-2-g365495a329 (Feb 20 2020 - 04:17:35 +0000) Voidlinux

Model: Pine64 Pinebook Pro
DRAM:  3.9 GiB
PMIC:  RK808 
MMC:   dwmmc@fe320000: 1, sdhci@fe330000: 0
In:    serial@ff1a0000
Out:   serial@ff1a0000
Err:   serial@ff1a0000
Model: Pine64 Pinebook Pro
## Error: Can't overwrite "serial#"
## Error inserting "serial#" variable, errno=1
rockchip_dnl_key_pressed: adc_channel_single_shot fail!
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
Card did not respond to voltage select!
starting USB...
Bus usb@fe380000: USB EHCI 1.00
Bus usb@fe3c0000: USB EHCI 1.00
Bus dwc3: Register 2000140 NbrPorts 2
Starting the controller
USB XHCI 1.10
Bus dwc3: Register 2000140 NbrPorts 2
Starting the controller
USB XHCI 1.10
scanning bus usb@fe380000 for devices... 1 USB Device(s) found
scanning bus usb@fe3c0000 for devices... 3 USB Device(s) found
scanning bus dwc3 for devices... 1 USB Device(s) found
scanning bus dwc3 for devices... 1 USB Device(s) found
       scanning usb for storage devices... 0 Storage Device(s) found

Device 0: unknown device
switch to partitions #0, OK
mmc0(part 0) is current device
Scanning mmc 0:1...
Found U-Boot script /boot.scr
865 bytes read in 13 ms (64.5 KiB/s)
## Executing script at 00500000
22583808 bytes read in 1005 ms (21.4 MiB/s)
63060 bytes read in 33 ms (1.8 MiB/s)
## Flattened Device Tree blob at 01f00000
   Booting using the fdt blob at 0x1f00000
Host not halted after 16000 microseconds.
ERROR: reserving fdt memory region failed (addr=0 size=0)
   Loading Device Tree to 00000000f5f0a000, end 00000000f5f1cfff ... OK

Starting kernel ...
```

The led turns green when it starts the kernel, and that's also where it starts displaying the rest of the log on the screen. I guess the kernel messages get relayed to serial on the other distro's?

Now, the kernel messages aren't necessarily a problem, but the last few unfortunately interfere with the login prompt.

One last thing of note: the messages on the screen don't always start appearing at the exact same time. Usually at around 1.7 seconds out of 3.9 in my case.

I wonder if this is the kernel or uboot doing this?

**Comment by houki0** [comment]
Also, login is a bit wonky. (Try pressing arrow keys.)
And the kernel messages can randomly interfere when you're in the shell.

**Comment by xtraeme** [comment]
The kernel cmdline should have `loglevel=4` like we do with GRUB.

**Comment by Skirmisher** [comment]
> When it boots, it first displays a red power light, which then turns green after a few seconds.

This is a feature of uboot—it's helpful for diagnosing boot issues, since the light turns green only after uboot passes control to the kernel. (If the kernel panics, it flashes red!)

**Comment by houki0** [comment]
It doesn't do that with the other distro's though, so I thought I'd report it. Seems pretty useful.

But I tried resizing the root partition a bit, but it doesn't boot anymore after it.
Uboot seems to work fine though. The light's green and the UART log is the exact same as ever.

Unfortunately, the kernel doesn't output to the UART, so I can't provide any logs this time, but it seems runit hasn't started yet. (acpid doesn't respond to short-pressing the power button.)

**Comment by Skirmisher** [comment]
> It doesn't do that with the other distro's though, so I thought I'd report it. Seems pretty useful

Right, I should have specified—this feature is in mainline uboot only. The 2017.09 Rockchip fork doesn't, and it's what many distros are using.

**Comment by houki0** [comment]
Apparently PARTUUID changes when resizing the partition.
After editing boot.scr through boot.txt with the new PARTUUID from `blkid` and compiling it with
`sudo mkimage -A arm -O linux -T script -C none -n "U-Boot boot script" -d boot.txt boot.scr`
it works again.

**Comment by renatoaguiar** [comment]
> Apparently PARTUUID changes when resizing the partition.
> After editing boot.scr through boot.txt with the new PARTUUID from `blkid` and compiling it with
> `sudo mkimage -A arm -O linux -T script -C none -n "U-Boot boot script" -d boot.txt boot.scr`
> it works again.

It didn't change when I tried. I think it depends on how you do it or what tool you use. In any case it is always good to run 'xbps-reconfigure -f pinebookpro-kernel' after changing things that could affect booting and that also runs 'mkimage', so you don't need to call it manually.

**Comment by renatoaguiar** [comment]
> The kernel cmdline should have `loglevel=4` like we do with GRUB.

https://github.com/void-linux/void-packages/pull/19380

**Comment by jcgruenhage** [comment]
@renatoaguiar I flashed the musl image to my eMMC, works nicely. Bonus: It seems to prefer the sdcard over the eMMC when booting! Did you change something there or were my previous attempts just failing because the boot itself failed somehow?

**Comment by renatoaguiar** [comment]
> @renatoaguiar I flashed the musl image to my eMMC, works nicely. Bonus: It seems to prefer the sdcard over the eMMC when booting! Did you change something there or were my previous attempts just failing because the boot itself failed somehow?

There is no other changes, so I guess boot itself was failing from eMMC.

**Comment by jcgruenhage** [comment]
Neither lightdm, nor gdm nor sway work on the latest image you posted, @renatoaguiar. I tried gdm in both xorg and wayland modes, lightdm with the gtk3 greeter with xorg and sway. For gdm, I couldn't find any error in the logs (just some warnings about systemd not being present), same with lightdm. i3 works, but I don't really want to have a system without graphical login. Did anything else change in the latest build? ftr, these are the packages I have installed, they should contain everything needed for this to work:
```
base-system-0.113_2
base-voidstrap-0.10_1
gnome-3.32.0_2
i3-4.18_1
lightdm-1.30.0_2
lightdm-gtk3-greeter-2.0.7_1
mesa-panfrost-dri-19.3.4_1
pinebookpro-base-0.1_1
sway-1.4_1
xorg-fonts-7.6_5
xorg-input-drivers-7.6_4
xorg-minimal-1.2_2
xtools-0.57_1
```

**Comment by jcgruenhage** [comment]
Thanks everyone for the work on this, btw. It's getting quite close now, you've been doing an awesome job!

**Comment by CameronNemo** [comment]
I think you should have `mesa-kmsro-dri` installed too, no?

**Comment by jcgruenhage** [comment]
Why would I need that? Trying now..

**Comment by CameronNemo** [comment]
just going based off https://panfrost.freedesktop.org/building-panfrost-mesa.html

**Comment by jcgruenhage** [comment]
So, installed that, sway still can't open a wayland socket and gdm still displays it "oh no, something went wrong" screen

**Comment by houki0** [comment]
That's weird. I did get sway to work on that image.

Slightly overkill list of relevant explicitly installed packages (basically all of them except user programs):
- base-system
- base-voidstrap
- elogind
- mesa
- mesa-dri
- pinebookpro-base
- polkit
- sway
- xdg-user-dirs
- xdg-utils
- and some stuff for testing a xfce/i3 setup.

About that i3 setup. Hovering over any xfce button, be it a launcher button or a application menu, results in a segfault. And `xinit i3` won't start without xterm, and it opens xterm as a i3 window that logs i3 itself, shutting down i3 when closed.
With default settings. I've never really messed around with xorg settings.

Most other DE's I tested end up crashing one way or another before showing anything usable. (KDE and MATE.)

It also seems [mesa-kmsro-dri](https://github.com/void-linux/void-packages/blob/master/srcpkgs/mesa-kmsro-dri), like many other mesa packages, is a dummy package linking to either mesa or mesa-dri.

**Comment by jcgruenhage** [comment]
that's very weird indeed. Currently not at home, I'll try to build a new image from scratch then and check whether it works then..

**Comment by jcgruenhage** [comment]
Still getting the same result with a freshly built image...

**Comment by jcgruenhage** [comment]
A few more details: I built the image by building a rootfs, then a platformfs, then the image using no special flags. The commit I built from was the one from #115. I installed a superset of the packages @houki0 mentioned. What makes this even weirder, on some image that I tried earlier, I could get it to work, but I don't remember which one.

**Comment by gogabr** [comment]
I am using the `void-pinebookpro-musl-20200221.img.xz` from @renatoaguiar for the third day now, and so far everything seems as smooth as possible. Lightdm, awesome, firefox, mpv all seem quite stable; sound works with `asound.conf` copied from Debian + sndiod.
Only problem was `texlive-bin` only available as glibc binaries and not working with gcompat; I had to manually copy  ld-linux-aarch64.so.1, libc and libm from a glibc system (luckily the names are different and do not conflict with musl). But that is not specific to pinebook pro.

**Comment by houki0** [comment]
It probably should be noted that my described sway setup worked on the glibc image.
The musl image I'm testing now gives a "Unable to create backendDRM device" error with another error underneath it on the same line that's not really clear.

Xfce does work perfectly with it though, and I'm currently resizing the partition so I can test KDE.

**Comment by houki0** [comment]
KDE works like a charm on musl!

**Comment by renatoaguiar** [comment]
> sound works with `asound.conf` copied from Debian + sndiod.

@gogabr Did you download asound.conf from some upstream repo or just copied from your running debian?

I'll include it in pinebookpro-base.

**Comment by gogabr** [comment]
I just copied it from my Debian partition.
I remember that some variants I downloaded did not work (maybe this? https://gitlab.manjaro.org/manjaro-arm/packages/community/pinebookpro-post-install/blob/master/asound.state)
I will test again later today when I come home.

**Comment by gogabr** [comment]
Checked this. The `asound.state` from Manjaro, linked above, works just fine.

**Comment by renatoaguiar** [comment]
Updated pinebookpro-base with audio and keyboard configuration from manjaro: https://github.com/void-linux/void-packages/pull/19564

**Comment by houki0** [comment]
I just found out  why my musl setup was software rendering.
I forgot to install mesa-dri.

Right now, sway works again, but I have the exact same issues with X11 as on glibc.
I think our panfrost is broken.

**Comment by anjandev** [comment]
Hello!

Excuse my ignorance here as I may have troubleshooted this issue incorrectly. I installed voidlinux on my pbp 2 weeks ago. Yesterday, I was getting ssl errors with xbps and spotty wifi. I suspected that it was an issue with the time. However, I had updated my time via `sudo ntpdate 0.ca.pool.ntp.org`.

I think this was finally fixed by setting
`HARDWARECLOCK="localtime"` in rc.conf and running `sudo ntpdate 0.ca.pool.ntp.org`.

The problem was with the hardware clock (which returned UTC time) I believe cause the system clock was correct (using localtime) and I was still getting ssl errors.

I dont know if this was the correct fix. If this was not the right fix, please let me know the correct way to fix so I can teach people how to fix this problem on the pinebook pro wiki.

**Comment by anjandev** [comment]
A word of warning for everyone, I cannot get libreoffice to work on the pinebook pro.

See:
https://github.com/void-linux/void-packages/issues/19642

Even flatpaks dont work. I was just able to reproduce this issue:
https://github.com/flathub/org.libreoffice.LibreOffice/issues/77

Please take a look at those issues if you can.

**Comment by xtraeme** [comment]
@anjandev I really doubt using `localtime` is a **fix**. Rather you should use a sw that sets up system time correctly, i.e `chrony`.

I do not know if this device has a real RTC, but these cheap devices normally don't.

**Comment by xtraeme** [comment]
@houki0 saying `panfrost` is broken with no logs is bullshit. Please paste logs to verify your words.

**Comment by houki0** [comment]
[Here's the most meaningful logs I could find](https://github.com/houki0/Logs/tree/master/xfce200302)

I came to that conclusion after seeing software rendering work, and panfrost crash, without fiddling with xorg settings first, so please take it with a grain of salt.

**Comment by renatoaguiar** [comment]
> I do not know if this device has a real RTC, but these cheap devices normally don't.

@xtraeme It does have a real RTC.

**Comment by q66** [comment]
@houki0 @xtraeme https://github.com/void-linux/void-mklive/issues/105#issuecomment-587171038

**Comment by houki0** [comment]
I forgot about that one.
Does manjaro use mesa-git?

**Comment by KeepBotting** [comment]
manjaro does use mesa-git, yes

https://gitlab.manjaro.org/manjaro-arm/packages/extra/mesa-git/-/blob/master/PKGBUILD

**Comment by BenEvolent333** [comment]
Using the musl build provided here, I can't use an X programs (I haven't checked Wayland) while mesa-dri is installed. I will try again after running updates, since I haven't updated in a while

**Comment by renatoaguiar** [comment]
Did anybody else notice an improvement in display support after upgrading to latest mesa version from official repository?

**Comment by houki0** [comment]
I hadn't noticed it updated yet.

But it seems to work flawlessly at first glance. The issues with XFCE I posted about are fixed.

**Comment by jcgruenhage** [comment]
> Did anybody else notice an improvement in display support after upgrading to latest mesa version from official repository?

Yes, this finally makes void usable on my pbp :)

There are still some quirks, but sway works, Firefox works, terminal emulators work sort-of (except for some bugs when doing tab completion, but that's probably a problem with my shell setup, even though the exact same setup works flawlessly on manjaro, and also on my workstation with void), and that's basically all I need :D

GDM still crashes, but that's okay, I'm using CDM in the mean time. Most Qt applications (tried pinentry-qt and pavucontrol-qt) seem to go a bit nuts, with mouse pointers not being where they are supposed to be and weird stuff like that. Has anyone else experienced weird behavior like that?

**Comment by BenEvolent333** [comment]
After installing the updated mesa-dri, I can open xorg, plasma, sddm, lightdm, lxqt and plasma (xorg, not wayland), but that's probably because I forgot to install something or other. However, there is a visual glitch on plasma where, when opening and closing applications, parts of the screen turn white.

**Comment by houki0** [comment]
I've been using it as a daily driver for the past few days, with a bare i3 setup with lightdm. 
Works pretty well until I enable a compositor, which drops the frames I get in glxgears to 55-45 fps or so, and makes the system a lot less responsive.

I haven't done any real testing with anything else yet, but from the little time I've spent testing KDE I can confirm I'm seeing the same issues as @BenEvolent333.

If I have the time I might try out an XFCE setup with its default window manager, to see if the problem is with compositing in general, or just with xcompmgr/picom.

But it actually is usable as a basic desktop OS now, so that's great! Thanks everyone!

**Comment by BenEvolent333** [comment]
@houki0 in my experience, Compton works just fine as a compositor, but the one included by default with KDE does not.

**Comment by sunnaryt** [comment]
it's not KDE's compositor but qt, in manjaro they have special "qt5-es2*" packages built for panfrost compatibility, but i actually have no idea what they are why they work

**Comment by jcgruenhage** [comment]
I've been running this for a while now, it's mostly working really good. Some applications don't work (that do work on manjaro), like mumble, but it's usable as a daily driver for me. @xtraeme, what do you think is still needed before a blog post and official images can be made available?

**Comment by jcgruenhage** [comment]
In case people haven't read the march update from the pine64 people: It currently looks like kernel 5.7 will include everything needed for the pbp, so the device specific kernel can be removed a lot sooner than most people probably expected.

**Comment by xtraeme** [comment]
@jcgruenhage I'm not an admin anymore nor I do have commit rights to push to www. You'll have to submit a news post to voidlinux.github.io and convince an admin to generate a new image for this device.

**Comment by KeepBotting** [comment]
until official images can be generated, are there any images up-to-date with this PR which can be written with `dd` to the pinebook pro's eMMC?

**Comment by jcgruenhage** [comment]
I built and shared this for some redditor a while ago https://cloudflare-ipfs.com/ipfs/QmTsWuKW9udVYYY8KfzA4AEQEwnAiptC8G1kUe56zsKiHK/void-pinebookpro-20200320.img.xz

**Comment by KeepBotting** [comment]
thanks, this boots up nicely from eMMC and also gives microSD priority if one is present

**Comment by jcgruenhage** [comment]
Yes. I think someone (probably me if no one else wants to chime in) has to write that blog post and ask the admins to generate official images.

**Comment by xtraeme** [comment]
well, someone is able to submit a PR. I'm not the one anymore! you can be the one

**Comment by KeepBotting** [comment]
after resizing my root partition i can no longer boot, is there a known workaround for this?

**Comment by jcgruenhage** [comment]
Depending on how you resize the partition and the filesystem in it, it might change the partitions UUID. Editing /etc/fstab should fix that. For me, resizing the partition with cfdisk and the fs with resize2fs worked without having to change the fstab.

**Comment by renatoaguiar** [comment]
I think we should automatically expand root partition on first boot or at least provide an easy CLI tool, e.g. "void-expand-rootfs", to do that properly (without changing UUID).

BTW, that should be a common issue for all targets. How are people dealing with it on Raspberry Pi, for example?

**Comment by jcgruenhage** [comment]
In general, how is using void as a daily driver working out of others here? I still have a bunch of issues which I couldn't be bothered to fix yet, such as audio not working, bluethooth audio stuttering terribly, Qt applications freezing immediately after launch and some other graphical tools just refusing to launch at all, like GDM. Getting Gnome to work would be amazing, even if it would likely be quite slow.

**Comment by CameronNemo** [comment]
* Built in audio works mostly great for me (have to disable speakers and enable speakers manually, headphone is always on)
* Have not tried bluetooth audio but bluetooth keyboard does work
* Qt applications work with latest mesa and the `qt5-wayland` package installed and the `QT_QPA_PLATFORM` set to `wayland` (although I am having some theming issues).
* No idea about GNOME.

**Comment by gogabr** [comment]
For me, mostly everything works. (Admittedly, my demands are not very high.) I built myself emacs, ghc and xmonad, so now I have my preferred setup. Java also works well enough to start Idea, but I haven't done much else with it.
For some reason, video does not play in Telegram.
I could not, so far, make suspend work.

**Comment by houki0** [comment]
All I'm missing is:
- Libreoffice (https://github.com/void-linux/void-packages/issues/19642)
- Hardware decoding/encoding (but that's not for Void to fix atm)
- My microphone doesn't work, but I haven't bothered to figure out why yet. (I'm experimenting with sndio)
- I have to manually enable the speakers in alsa like @CameronNemo.

Don't know about bluetooth audio, but my bluetooth mouse works fine.
I'm not having any problems with qt so far.

**Comment by thallian** [comment]
I am running it on musl with sway and most stuff works nicely (doing full disk encryption too). Using pulseaudio.

- speakers were very low volume in the beginning, had to up the volume in alsamixer, now they are fine
- bluetooth audio works fine
- alacritty/kitty don't work because of missing gles 3 support (that should change, already experimental support for that in mesa)
- rustup doesn't seem to have support for aarch64-musl (rustc and cargo are missing), you need to compile stuff like rust-analyzer yourself
- the trackpad is not too bad but I am having trouble doing small, precise movements. Maybe I just need to get used to it more.

**Comment by jcgruenhage** [comment]
@thallian How have you set up FDE? I've messed with void-mklive a bit in that direction, but couldn't get it to boot so far.

**Comment by thallian** [comment]
@jcgruenhage I used void-mklive unchanged to build the image. Did an lvm on luks installation with boot unencrypted on a seperate partition. 

In /boot/boot.txt you need to add/update some bootargs:
- `root=/dev/void/root` (void is the name of my lvm volume group and root is the name of the volume)
- `rd.auto=1` (to enable auto assembly of luks/lvm, without that it wont ask for a password and won't boot)
- `cryptdevice=/dev/mmcblk2p2:lvm` (you can use persistent block naming too here, should be the better choice)

Update /boot/boot.scr: `mkimage -A arm -O linux -T script -C none -n "boot script" -d /boot/boot.txt /boot/boot.scr`

Build initramfs (dracut should pick up on the needed modules automatically): `dracut`

Convert the initramfs to a the u-boot format: 
- `mkimage -A arm64 -O linux -T ramdisk -C none -d /boot/initRamfs-5.6.0_1.img initRamfs-5.6.0_1.img`
-`mv initRamfs-5.6.0_1.img /boot/initRamfs-5.6.0_1.img`

The boot.txt provided by pinebookpro-uboot picks that up automatically.

Add the cryptdevice to `/etc/crypttab`: `crypt /dev/mmcblk2p2 none luks` (again, uuid should work too).

That's what I did, hope it helps. I am also not sure if I can leave parts of that out, still trying stuff and I am very new to the arm world :)

**Comment by Skirmisher** [comment]
Anyone else having issues with setting up for external modules in the latest kernel versions? I haven't tried since 5.5, but I'm getting this when running `make modules_prepare` in `/usr/src/kernel-headers-5.7.0_1`:

```
arch/arm64/kernel/vdso/Makefile:12: lib/vdso/Makefile: No such file or directory
make[1]: *** No rule to make target 'lib/vdso/Makefile'.  Stop.
make: *** [arch/arm64/Makefile:173: vdso_prepare] Error 2
```

**Comment by jcgruenhage** [comment]
Anyone else having problems with bluetooth audio stuttering?

When booting, it complains about a `clm_blob` not being available, so the device may have limited channels, which sounds related.

**Comment by KeepBotting** [comment]
I'll help test, would like to switch my eMMC to Void with upstream kernel/u-boot :)

**Comment by jcgruenhage** [comment]
Thanks for the offer, but I'm not sure how much testing is actually needed/useful there. It either boots, or it doesn't, and for the "it doesn't boot" case there's not much one can do without the UART cable (which I have, so debugging won't be a problem). Still, once I have working builds, I'll publish them here.

**Comment by valadaa48** [comment]
Just got my PBP today. I tried running:

    sudo ./mkrootfs.sh -o aarch64latest aarch64.
    sudo ./mkplatformfs.sh pinebookpro aarch64latest
    sudo ./mkimage.sh void-pinebookpro-PLATFORMFS-20200216.tar.xz

then unxz the resulting img.xz then dd'd to a microsd. When I insert the microsd and power on the power LED flashes between green and red. I did a `sudo pacman -Syu` in manjaro before trying as well as I read that some packages may need upgrading to boot off of the microsd.

Any ideas what I might be doing wrong?

**Comment by jcgruenhage** [comment]
You might need to manually flash the new uboot after updating the uboot package. https://gitlab.manjaro.org/manjaro-arm/packages/core/uboot-pinebookpro/-/raw/master/uboot-pinebookpro.install has the instructions for that

**Comment by CameronNemo** [comment]
>Manjaro switched from a custom kernel to the upstream kernel, so we shouldn't need a pinebook pro specific kernel anymore.

Their "upstream kernel" has tons of patches... many of them PBP related. https://gitlab.manjaro.org/manjaro-arm/packages/core/linux

**Comment by jcgruenhage** [comment]
Ah right, didn't see that. 19 patches in total is indeed wuite a lot. So that means we'll stay on a custom kernel package on the pinebook pro for now. I hope that the people who wrote those patches are contributing them upstream..

**Comment by CameronNemo** [comment]
It seems like they are decidedly not. I took a look at some of the larger sets and they all have "do not mainline" and "hack" stated in them. One of the patches had a call to `usleep_range` in it haha. They mainly accomplish:

* deep suspend (states it is a hack, not for mainline, and is also reportedly not quite functional with the newer blob-free u-boot)
* USB type-c altmode (uses extcon, which [the rockchip subsystem maintainer said yesterday is a hack that should not be extended](http://lists.infradead.org/pipermail/linux-rockchip/2020-September/022148.html))
* A set of patches to power cycle the bluetooth and WiFi chips on reboots to ensure that the firmware does not act up. *might* be upstream-able, but it is a longshot. This is the patchset that contains the sleep, and it also seems not at all specific to the BT/WiFi chipsets in use on the PBP. Maybe we can find a way to manually power cycle the devices from userspace during boot?
* some ALSA/sound fixes, these may be upstream-able. I would give it about a 50/50 chance.
* some small DTS tweaks for the USB type-c port role and enabling earlycon. Not sure why the latter is useful, but they both seem entirely reasonable to upstream.

Overall, if we are fine sacrificing deep suspend and USB-C altmode until someone finds solutions for both using the right methods (ATF's PSCI interface and the "battle plan" in Heiko's mail linked above, respectively), then I would say we have a highly maintainable kernel. BT/WiFI stability is non-negotiable for me, but that patchset is the smallest of the first three large ones.

**Comment by CameronNemo** [comment]
Hi all, I have a PR to update to U-Boot to the 2020.10 release: https://github.com/void-linux/void-packages/pull/21199

**Comment by Algorithmus** [comment]
I had an issue with some firmware blobs missing, so the wifi didn't work correctly when I ran the voidlinux image on my pinebook pro. The firmware came from here: https://gitlab.manjaro.org/manjaro-arm/packages/community/ap6256-firmware

**Comment by snufkin99** [comment]
> @jcgruenhage I used void-mklive unchanged to build the image. Did an lvm on luks installation with boot unencrypted on a seperate partition.
> 
> In /boot/boot.txt you need to add/update some bootargs:
> 
>     * `root=/dev/void/root` (void is the name of my lvm volume group and root is the name of the volume)
> 
>     * `rd.auto=1` (to enable auto assembly of luks/lvm, without that it wont ask for a password and won't boot)
> 
>     * `cryptdevice=/dev/mmcblk2p2:lvm` (you can use persistent block naming too here, should be the better choice)
> 
> 
> Update /boot/boot.scr: `mkimage -A arm -O linux -T script -C none -n "boot script" -d /boot/boot.txt /boot/boot.scr`
> 
> Build initramfs (dracut should pick up on the needed modules automatically): `dracut`
> 
> Convert the initramfs to a the u-boot format:
> 
>     * `mkimage -A arm64 -O linux -T ramdisk -C none -d /boot/initRamfs-5.6.0_1.img initRamfs-5.6.0_1.img`
>       -`mv initRamfs-5.6.0_1.img /boot/initRamfs-5.6.0_1.img`
> 
> 
> The boot.txt provided by pinebookpro-uboot picks that up automatically.
> 
> Add the cryptdevice to `/etc/crypttab`: `crypt /dev/mmcblk2p2 none luks` (again, uuid should work too).
> 
> That's what I did, hope it helps. I am also not sure if I can leave parts of that out, still trying stuff and I am very new to the arm world :)
> 
> EDIT: so, the dracut part was only necessary because it failed during the initial installation (don't know why), after running `xbps-reconfigure -f pinebookpro-kernel` it generated the initramfs correctly.
> 
> And the crypttab part is not necessary either, only makes you input the password twice.

Awesome work! Thanks to everyone involved in this project. I really enjoy the Pinebook Pro and Void Linux is one of my favorite distros! I have a question regarding fde. I know how to install Void on x86 with lvm/fde but I just can't get it to work on the Pinebook Pro. I've tried moving files from the image after setting up lvm with unencrypted boot following the steps mentioned. I've tried so many combinations and I'm lost. I feel like I'm missing something important in this little "guide". Would you or someone else mind explaining this a little more in depth? Like a small howto would be very appreciated. Sorry if I'm missing something obvious here but I think this is also a great way to learn. Thnx and take care all!

**Comment by thallian** [comment]
@snufkin99 
You can take a look at how I am doing it when installing: https://code.vanwa.ch/sebastian/void-linux-installer

The relevant parts for fde are in the luks, fstab and boot_scr functions (just look for these words, the function names are not exactly that). Also ensure to have the packages dracut, lvm2 (if using lvm) and cryptsetup installed.

My email is on the website that is linked on my gitlab profile, that might be a better way to figure out communications without spamming the issue here : )

**Comment by the-maldridge** [comment]
@jcgruenhage 12/12 are done on this ticket, are there remaining work items you'd like it kept open for?

**Comment by jcgruenhage** [comment]
@the-maldridge yeah, sorry, haven't updated this ticket in a while. Ideally I'd like to see images on the download site, added that to the list. Does that sound okay to you, or would you prefer having the issue closed as it is right now?

**Comment by via** [comment]
I'm experiencing kernel oops's in the pinebookpro-kernel 5.10.35 that prevent void from being usable. They usually trigger when the cpu/disk is under load, and I've reproduced this on both a usb root device as well as the emmc. Has anyone else seen this?

```
[  601.089721] Unable to handle kernel paging request at virtual address ffff00003d530000
[  601.090420] Mem abort info:
[  601.090667]   ESR = 0x96000047
[  601.090937]   EC = 0x25: DABT (current EL), IL = 32 bits
[  601.091401]   SET = 0, FnV = 0
[  601.091670]   EA = 0, S1PTW = 0
[  601.091947] Data abort info:
[  601.092200]   ISV = 0, ISS = 0x00000047
[  601.092537]   CM = 0, WnR = 1
[  601.092799] swapper pgtable: 4k pages, 48-bit VAs, pgdp=0000000001815000
[  601.093382] [ffff00003d530000] pgd=00000000f3ff9003, p4d=00000000f3ff9003, pud=00000000f3ff8003, pmd=00000000f3e18003, pte=006800000000446e
[  601.094479] Internal error: Oops: 96000047 [#1] SMP
[  601.094906] Modules linked in: clk_rk808 rk808_regulator rtc_rk808 rk808 cw2015_battery fan53555 regmap_i2c gpio_keys hid_multitouch panfrost gpu_sched rockchipdrm dw_mipi_dsi dw_hdmi analogix_dp cec drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops i2c_rk3x panel_simple drm drm_panel_orientation_quirks i2c_core pwm_bl btrfs blake2b_generic xor xor_neon zstd_compress raid6_pq
[  601.097925] CPU: 4 PID: 5259 Comm: gzip Not tainted 5.10.35_1 #1
[  601.098448] Hardware name: Pine64 Pinebook Pro (DT)
[  601.098876] pstate: 20000005 (nzCv daif -PAN -UAO -TCO BTYPE=--)
[  601.099405] pc : clear_page+0x14/0x28
[  601.099731] lr : kernel_init_free_pages+0x68/0xa0
[  601.100141] sp : ffff800018c839e0
[  601.100432] x29: ffff800018c839e0 x28: ffff8000133fb6c0 
[  601.100898] x27: ffff800018c83b10 x26: 0000000000000000 
[  601.101364] x25: 0000000000000000 x24: ffff8000133fcfc0 
[  601.101829] x23: 0000000000000001 x22: ffff000000000000 
[  601.102294] x21: 0000000000f54c40 x20: ffff00002ae08e80 
[  601.102759] x19: 0000000000f54c00 x18: 0000000000000000 
[  601.103224] x17: 0000000000000000 x16: 0000000000000000 
[  601.103689] x15: 0000000000000000 x14: 0000000000000003 
[  601.104154] x13: 00000000000b67d4 x12: 0000000000000040 
[  601.104619] x11: ffff00000f2e8248 x10: ffff800018c83d50 
[  601.105084] x9 : 0000000000001000 x8 : 00000000f7e00000 
[  601.105549] x7 : 0000000000000018 x6 : ffff800013c71870 
[  601.106014] x5 : ffff800013c71870 x4 : 0000000000000001 
[  601.106479] x3 : 0000000000000881 x2 : 0000000000000004 
[  601.106944] x1 : 0000000000000040 x0 : ffff00003d530000 
[  601.107410] Call trace:
[  601.107627]  clear_page+0x14/0x28
[  601.107919]  prep_new_page+0xa8/0xd4
[  601.108234]  get_page_from_freelist+0x1e0/0x280
[  601.108630]  __alloc_pages_nodemask+0x138/0x2c0
[  601.109028]  pagecache_get_page+0x1dc/0x3ec
[  601.109394]  grab_cache_page_write_begin+0x28/0x4c
[  601.109815]  ext4_da_write_begin+0xb4/0x3d0
[  601.110183]  generic_perform_write+0xa8/0x1d0
[  601.110564]  ext4_buffered_write_iter+0x9c/0x180
[  601.110968]  ext4_file_write_iter+0x34/0x60
[  601.111335]  new_sync_write+0xe8/0x184
[  601.111662]  vfs_write+0x220/0x2b4
[  601.111961]  ksys_write+0x68/0xf4
[  601.112252]  __arm64_sys_write+0x20/0x2c
[  601.112602]  el0_svc_common.constprop.0+0x78/0x1a0
[  601.113021]  do_el0_svc+0x24/0x90
[  601.113314]  el0_svc+0x14/0x20
[  601.113584]  el0_sync_handler+0x1a4/0x1b0
[  601.113935]  el0_sync+0x178/0x180
[  601.114229] Code: d53b00e1 12000c21 d2800082 9ac12041 (d50b7420) 
[  601.114762] ---[ end trace 3ed49acd6eab72af ]---
[  601.286828] Unable to handle kernel paging request at virtual address ffff00003d5b0000
[  601.287680] Mem abort info:
[  601.287944]   ESR = 0x96000047
[  601.288219]   EC = 0x25: DABT (current EL), IL = 32 bits
[  601.288687]   SET = 0, FnV = 0
[  601.288960]   EA = 0, S1PTW = 0
[  601.289237] Data abort info:
[  601.289494]   ISV = 0, ISS = 0x00000047
[  601.289834]   CM = 0, WnR = 1
[  601.290102] swapper pgtable: 4k pages, 48-bit VAs, pgdp=0000000001815000
[  601.290690] [ffff00003d5b0000] pgd=00000000f3ff9003, p4d=00000000f3ff9003, pud=00000000f3ff8003, pmd=00000000f3e18003, pte=00680000000044ae
[  601.291804] Internal error: Oops: 96000047 [#2] SMP
[  601.292236] Modules linked in: clk_rk808 rk808_regulator rtc_rk808 rk808 cw2015_battery fan53555 regmap_i2c gpio_keys hid_multitouch panfrost gpu_sched rockchipdrm dw_mipi_dsi dw_hdmi analogix_dp cec drm_kms_helper syscopyarea sysfillrect sysimgblt fb_sys_fops i2c_rk3x panel_simple drm drm_panel_orientation_quirks i2c_core pwm_bl btrfs blake2b_generic xor xor_neon zstd_compress raid6_pq
[  601.295320] CPU: 0 PID: 5279 Comm: htop Tainted: G      D           5.10.35_1 #1
[  601.295967] Hardware name: Pine64 Pinebook Pro (DT)
[  601.296400] pstate: 20000005 (nzCv daif -PAN -UAO -TCO BTYPE=--)
[  601.296938] pc : clear_page+0x14/0x28
[  601.297270] lr : kernel_init_free_pages+0x68/0xa0
[  601.297684] sp : ffff800018d43b30
[  601.297979] x29: ffff800018d43b30 x28: ffff8000133fb6c0 
[  601.298453] x27: ffff800018d43c60 x26: 0000000000000000 
[  601.298927] x25: 0000000000000000 x24: ffff8000133fcfc0 
[  601.299400] x23: 0000000000000001 x22: ffff000000000000 
[  601.299873] x21: 0000000000f56c40 x20: ffff0000f2ef6580 
[  601.300346] x19: 0000000000f56c00 x18: 0000000000000000 
[  601.300818] x17: 0000000000000000 x16: 0000000000000000 
[  601.301291] x15: 0000000000000000 x14: 0000000000000003 
[  601.301764] x13: 00000000000b66d8 x12: 0000000000000000 
[  601.302237] x11: 0000000000000000 x10: ffff00002dcc1000 
[  601.302710] x9 : 0000000000000001 x8 : 00000000f7e00000 
[  601.303183] x7 : 0000000000000018 x6 : ffff800013c71870 
[  601.303656] x5 : ffff800013c71870 x4 : 0000000000000001 
[  601.304128] x3 : 0000000000000881 x2 : 0000000000000004 
[  601.304601] x1 : 0000000000000040 x0 : ffff00003d5b0000 
[  601.305074] Call trace:
[  601.305297]  clear_page+0x14/0x28
[  601.305596]  prep_new_page+0xa8/0xd4
[  601.305918]  get_page_from_freelist+0x1e0/0x280
[  601.306322]  __alloc_pages_nodemask+0x138/0x2c0
[  601.306725]  do_cow_fault+0x3c/0x220
[  601.307043]  do_fault+0x3c/0x250
[  601.307333]  handle_pte_fault+0x74/0x1a0
[  601.307684]  __handle_mm_fault+0x10c/0x190
[  601.308049]  handle_mm_fault+0x9c/0x200
[  601.308395]  do_page_fault+0x168/0x3e4
[  601.308732]  do_translation_fault+0xb0/0xdc
[  601.309106]  do_mem_abort+0x44/0xa4
[  601.309418]  el0_da+0x28/0x34
[  601.309686]  el0_sync_handler+0x168/0x1b0
[  601.310043]  el0_sync+0x178/0x180
[  601.310347] Code: d53b00e1 12000c21 d2800082 9ac12041 (d50b7420) 
[  601.310886] ---[ end trace 3ed49acd6eab72b0 ]---

```

**Comment by via** [comment]
I still was using u-boot from my previous debian installation.  Using the void package's u-boot on my emmc seems to have resolved this.

**Comment by vdo** [comment]
Are the final images going to be generated?

**Comment by Johnnynator** [comment]
> Are the final images going to be generated?

I wanted to wait until uboot has a version released that supports initialising the display panel and allow for some crude boot menu. Afaik 2021.07 should support that. (And maybe also fixup or default non pinebook pro kernel config to work fine on a pbp). 

So maybe August, maybe a bit later.

**Comment by CameronNemo** [comment]
@Johnnynator have you by chance tried a U-Boot with ATF with the suspend patch? (https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/9616) I have been meaning to give it a spin, but have not had much time.

**Comment by jcgruenhage** [comment]
Re uboot: we should pribably fix that it breaks customized uboot stuff on each update: my luks encrypted pbp refuses to boot because I forgot to fix that after an upgrade. Easy-ish to fix, but I'll have to whip out a live system for that again

**Comment by Johnnynator** [comment]
Yeah, hardcoding the bootargs in the uboot kernel hook is bad. I currently use u-boot-menu, which just reuses the bootargs if the booted system. Should probably also just noticed, why do we even hardcode the Mac address in the uboot image

**Comment by CameronNemo** [comment]
I would be glad to see void just ship u-boot-menu by default for the PBP. Much cleaner, easier to understand and customize.

**Comment by Skirmisher** [comment]
If the latest u-boot has display support, then using u-boot-menu is probably the best thing to do. I do remember an oddity with u-boot-menu's extlinux.conf support where the timeout value is evaluated as tenths of a second for some reason, and if the value was less than 10 then it was interpreted as "no timeout", meaning your computer was effectively hung if you had no display or serial console to proceed with.

Also, last I checked, the install.d script that the u-boot-menu package uses some broken logic for figuring out the boot and root filesystems, and uses PARTUUID instead of UUID on the kernel cmdline, which is dubious. I have a fixed version of the script, but I haven't used the PBP in a long time so I haven't updated the package.

**Comment by Johnnynator** [comment]
Yeah, the u-boot-menu script would need some more work. (I only ever fixed it to not break on dash as sh).

As alternative using grub or refind might even give a more user friendly (and familiar) interface, and are more tested than half broken extlinux generators named u-boot-menu :).
But my last attempt with getting an image with refind resulted in it not finding any systems, and a ton of errors spammed in u-boot (too fast to read, need to check it over serial).

And yes 2021.07 has working display support, but I need to investigate where my builds (also with 2021.04) fail to boot our kernels,

**Comment by CameronNemo** [comment]
>Also, last I checked, the install.d script in the u-boot-menu package uses some broken logic for figuring out the boot and root filesystems, and uses PARTUUID instead of UUID on the kernel cmdline, which is dubious.

```
➜ rg PARTUUID srcpkgs/u-boot-menu/
➜ rg UUID srcpkgs/u-boot-menu/
➜ rg root srcpkgs/u-boot-menu/
```

Not seeing any handling for root at all. Looks like you inherited it from the PBP specific kernel script:

```
➜ rg PARTUUID srcpkgs/pinebookpro-uboot/
srcpkgs/pinebookpro-uboot/files/kernel.d/uboot
6:partuuid=$(blkid -o value -s PARTUUID "${dev}")
17:setenv bootargs console=ttyS2,1500000 console=tty1 root=PARTUUID=${partuuid} rootwait video=eDP-1:1920x1080@60 loglevel=4
```

u-boot-menu does have some weird logic for /boot, but the logic seems to work. I guess it could be simplified a bit (e.g. just check if `mountpoint -q /boot` and prefix the paths with `/boot` otherwise).

>I have a fixed version of the script, but I haven't used the PBP in a long time so I haven't updated the package.

Well feel free to shoot your changes over. I would be curious about what is broken. I don't doubt that there are edge cases not covered, but for the images created by void-mklive it seems to do alright.

**Comment by Skirmisher** [comment]
> Not seeing any handling for root at all. Looks like you inherited it from the PBP specific kernel script

Ah you're right, I got that mixed up sometime in the last while. I'll take a look at all the scripts again when I pull out my PBP again.

**Comment by user18130814200115-2** [comment]
As for the uboot supporting visual boot and suspend to ram, you might consider using [tow boot](https://github.com/Tow-Boot/Tow-Boot/).

I have been using this fork on my pinebookPro for a while and it works well.
It is also the uboot used for the [unofficial archlinuxARM repo](https://github.com/SvenKiljan/archlinuxarm-pbp)

**Comment by CameronNemo** [comment]
@user18130814200115-2 have you confirmed that it works with suspend to RAM? I have tried to get s2ram working and failed.

**Comment by user18130814200115-2** [comment]
I managed to get it working a while back on ArchlinuxARM with manjaro 5.13.? kernel.
I stopped using it though because it does not work with NVME.

The visual boot works great though, it displays useful information and has a bootmenu which can also boot from usb. Generally an improvement on das uboot IMO
